### ISSUE TITLE
High levels of open file descriptors while uploading a zip with lots of files in it

### ISSUE THREAD

**Issue by PaulBoon** [IQSS/dataverse#6723]
We are monitoring the number of open file descriptors ever since we had serious problems with it in the past (last year).
When the number of open file descriptors of the glassfish process is 3000 or higher (max is 4096) I get a mail, this happened several times in the past days. Now, the good news is that it goes down to the 'basis' value of around 700.
But if the peak will be too high, the system will fail and I don't want that to happen, besides this I get a lot of warning emails from the ‘monitoring’ script.
 
A typical run looks like the following:
 
2020-02-27T15:26:01+0100,742
2020-02-27T15:27:02+0100,1744
2020-02-27T15:28:01+0100,1744
2020-02-27T15:29:01+0100,2747
2020-02-27T15:30:01+0100,2747
2020-02-27T15:31:01+0100,1450
2020-02-27T15:32:01+0100,1452
2020-02-27T15:33:01+0100,2453
2020-02-27T15:34:01+0100,3456
 
With each line the timestamp followed by the number of open file descriptors, increments in steps of ~1000 (sampling every minute). 

Below a chart showing what happened on that day:
![image](https://user-images.githubusercontent.com/2151103/75879150-2b314a80-5e1b-11ea-84a2-bf45d1ccc8e3.png)
.

After another warning mail, we checked the logs

`[2020-03-02T13:51:04.422+0100] [glassfish 4.1] [SEVERE] [] [edu.harvard.hul.ois.jhove] [tid: _ThreadID=52 _ThreadName=jk-connector(3)] [timeMillis: 1583153464422] [levelValue: 1000] [[
  Testing SEVERE level]]
 
[2020-03-02T13:51:04.431+0100] [glassfish 4.1] [SEVERE] [] [edu.harvard.hul.ois.jhove] [tid: _ThreadID=52 _ThreadName=jk-connector(3)] [timeMillis: 1583153464431] [levelValue: 1000] [[
  Testing SEVERE level]]
 
[2020-03-02T13:51:04.437+0100] [glassfish 4.1] [SEVERE] [] [edu.harvard.hul.ois.jhove] [tid: _ThreadID=52 _ThreadName=jk-connector(3)] [timeMillis: 1583153464437] [levelValue: 1000] [[
  Testing SEVERE level]]
 
[2020-03-02T13:51:04.447+0100] [glassfish 4.1] [SEVERE] [] [edu.harvard.hul.ois.jhove] [tid: _ThreadID=52 _ThreadName=jk-connector(3)] [timeMillis: 1583153464447] [levelValue: 1000] [[
  Testing SEVERE level]]
 
[2020-03-02T13:51:04.450+0100] [glassfish 4.1] [WARNING] [] [edu.harvard.iq.dataverse.util.FileUtil] [tid: _ThreadID=52 _ThreadName=jk-connector(3)] [timeMillis: 1583153464450] [levelValue: 900] [[
  Zip upload - too many files.]]
 
[2020-03-02T13:51:04.450+0100] [glassfish 4.1] [WARNING] [] [edu.harvard.iq.dataverse.util.FileUtil] [tid: _ThreadID=52 _ThreadName=jk-connector(3)] [timeMillis: 1583153464450] [levelValue: 900] [[
  Unzipping failed; rolling back to saving the file as is.]]
 `

This is from the code in: edu.harvard.iq.dataverse.util.FileUtil
There we see that there is a limit in: 
edu/harvard/iq/dataverse/util/SystemConfig.java

`/**
     * The default number of datafiles that we allow to be created through 
     * zip file upload.
     */
    private static final int defaultZipUploadFilesLimit = 1000; 
    private static final int defaultMultipleUploadFilesLimit = 1000;
    private static final int defaultLoginSessionTimeout = 480; // = 8 hours
 `
It is not set by us, so it should indeed be the default 1000. 

Next I tested uploading a zip with a lot of files >1000 and several things happened. 
On the GUI I got the message: "The number of files in the zip archive is over the limit (1000); please upload a zip archive with fewer files, if you want them to be ingested as individual DataFiles."
The files in the zip where indeed not extracted into the dataset, but the complete zip was added, as is. 

But the following problems occurred: 
- The number of open file descriptors peeked. 
- All the temporary files where left in the Dataverse temp dir. 

It would be much better if those two things did not happen.

Furthermore, I tested it with a special constructed zip with exactly 1001 files in it. 
Then all the files where extracted, although 1001 > 1000. 

If I change the limit to 999, with: 
'curl -X PUT -d '999' http://localhost:8080/api/admin/settings/:ZipUploadFilesLimit'
The zip is added 'as is'. This looks like a off-by-one error, which would be nice to have fixed in order to minimise confusion. 
[1001_files.zip](https://github.com/IQSS/dataverse/files/4287255/1001_files.zip)

**Comment by poikilotherm** [comment]
This might be related to my findings in #6656 about file descriptors not being closed properly after upload.

**Comment by pdurbin** [comment]
> [1001_files.zip](https://github.com/IQSS/dataverse/files/4287255/1001_files.zip)

@PaulBoon thanks for attaching a test zip and the detailed explanation. This performance issue probably didn't fix itself, right? 😄  Have you tested lately? I should try it myself sometime.

**Comment by PaulBoon** [comment]
Hi @pdurbin, No we have not tested this for a long time now. It might be that much of the resource leakage was fixed by fixing https://github.com/IQSS/dataverse/issues/6656.

**Comment by mreekie** [comment]
Grooming:
- this sounds like we're sizing a reproduction of an issue so maybe this is a spike?
- needs: @jggautier or @sbarbosadataverse @kcondon @pdurbin

**Comment by mreekie** [comment]
Sizing:
 - Talked through this with Kevin.
 - It looks like this is not too hard to reproduce.
 - This looks to be about a 3
 - This is a spike:

**Comment by kcondon** [comment]
From the initial description, there seems to be 3 concerns:
1. DANS has a sys admin alert for when file descriptors rise past a certain level (3000). It has happened several times but then subsides to a base of 700. They were concerned that it could grow to the limit of 4096 and then fail. So want to understand and control this. In one case it was associated with a zip file containing too many files, ie. > 1000.
[Kevin] I've tested this with a 1000 file zip and it just looks like the system briefly opens a file descriptor for every unpacked file in the zip. I think this is normal operation?
2. When testing with a zip > 1000 files, open file descriptors again peaked and all the files were left in Dataverse temp dir. He would prefer that neither thing happened if it was not a valid upload.
[Kevin] This is the way upload works -we do not know how many files are in the zip file before unzipping it, at which point the descriptors are opened and with the prime faces upload widget, abandoned upload files are orphaned in temp so you would need a cleaner job to check that directory in general.
3. The failure message in the ui is: "The number of files in the zip archive is over the limit (1000)..." but the actual limit is 1001, not 1000 so there is an off by one error. This is confusing.
[Kevin] I would need to confirm this.

**Comment by qqmyers** [comment]
FWIW: It is very easy to look inside a zip and get the file count, so we should be able to avoid making temp files when you're over the limit.

**Comment by landreev** [comment]
@qqmyers 
> FWIW: It is very easy to look inside a zip and get the file count, so we should be able to avoid making temp files when you're over the limit.

Yeah, that was on account of my ignorance: back when I implemented this originally, I did not know that zip archives had a directory at the end of the file ☹️. I took a quick look at the methods in ZipInputStream and assumed zip was like tar, and you just had to read through the entire byte stream to get the list. And I didn't remember to go back and fix it once I did learn how zips worked. 😱 

I just put together some extra code, to make a quick check on the files directory, against both the number of files limit and the individual size limit. This, I believe, should make everything else mentioned in this issue more manageable; until we have more time to invest in this. (Or until we retire auto-unzipping altogether - I'm serious).
